forcetree.c

```diff
@@ -221,152 +313,143 @@ int force_tree_build_single(int npart)
-             * We create a new internal node with empty subnodes at the end of the array, and
-             * use it to replace the particle in the parent's subnode.
-             * Then we loop back in the hopes that we can add the particle we are currently working on
-             * to the new node.*/
-            /*Atomically add an extra node*/
-            if(numfree_thread == 0){
-                nfree_thread = atomic_fetch_and_add(&nfree, NODECACHE_SIZE);
-                numfree_thread = NODECACHE_SIZE;
             }
-            const int ninsert = nfree_thread++;
-            numfree_thread--;
-            /*If we already have too many nodes, exit loop.*/
-            if(nfree_thread >= All.MaxPart + MaxNodes)
-            {
-/*                 message(1, "maximum number %d of tree-nodes reached for particle %d.\n", MaxNodes, i); */
+        }
+        while(child >= firstnode);
+
 #ifdef OPENMP_USE_SPINLOCK
-                pthread_spin_unlock(&SpinLocks[this-All.MaxPart]);
+        /*Now lock this node.*/
+        pthread_spin_lock(&SpinLocks[this-firstnode]);
 #endif
-                break;
-            }
-            struct NODE *nfreep = &Nodes[ninsert];	/* select desired node */
-            int j;
-            const MyFloat lenhalf = 0.25 * Nodes[this].len;
-
-            nfreep->len = 0.5 * Nodes[this].len;
-
-            for(j = 0; j < 3; j++) {
-                /* Detect which quadrant we are in by testing the bits of subnode:
-                 * if (subnode & [1,2,4]) is true we add lenhalf, otherwise subtract lenhalf*/
-                const int sign = (subnode & (1 << j)) ? 1 : -1;
-                nfreep->center[j] = Nodes[this].center[j] + sign*lenhalf;
-            }
-            for(j = 0; j < 8; j++)
-                nfreep->u.suns[j] = -1;
-            nfreep->hmax = 0;
-
-            /*Re-add the particle to the new internal node*/
-            const int child_subnode = get_subnode(nfreep, this, child, shift -3, MortonKey);
-
-            nfreep->u.suns[child_subnode] = child;
-
-            /* Mark this node in the parent: this goes last
-             * so that we don't access the child before it is constructed.*/
-            Nodes[this].u.suns[subnode] = ninsert;
 
-            /* We are now done adding data to this node,
-             * so we can move the SpinLock to the child.*/
+        /*Check nothing changed when we took the lock*/
+        #pragma omp atomic read
+        child = Nodes[this].u.suns[subnode];
+        /*If it did, walk again*/
+        while(child >= firstnode)
+        {
 #ifdef OPENMP_USE_SPINLOCK
-            /*Initialise the SpinLock off*/
-            pthread_spin_init(&SpinLocks[ninsert-All.MaxPart], 0);
-            /*Unlock the parent*/
-            pthread_spin_unlock(&SpinLocks[this-All.MaxPart]);
+            /*Move the lock to the child*/
+            pthread_spin_lock(&SpinLocks[child-firstnode]);
```

Comment:
I think you may have to put this and the check above into a loop.

Is there a way to prove the child node is not split again before the lock is grabbed in the second try?

Reply:
If this is made into a loop it is probably not very difficult to revert to hashing of child when the node is sufficiently small.  If we do it here we avoid additional splitting for each colliding particle.

Reply:
I don't completely understand what you mean here: this is in a loop already. If the child is split we simply move the lock one more level down. The child cannot be split while we hold the lock on it, and that is the reason we lock the child before unlocking the parent: we always have a lock on something.

Reply:
I worry the same condition requirement the second check may require a third check. The child cannot be split while we hold the lock on it, but the child of the child (which becomes the new child) may.

Reply:
But in that case we will just keep walking until we find something that we can take the lock on. We always take the lock on a node before checking the node (this part is actually identical to the earlier code: the change is the addition of a lockless fast path earlier).